File: src/BreakfastProvider.Api/Controllers/DailySpecialsController.cs
using BreakfastProvider.Api.Models.Requests;
using BreakfastProvider.Api.Models.Responses;
using BreakfastProvider.Api.Services;
using Microsoft.AspNetCore.Mvc;

namespace BreakfastProvider.Api.Controllers;

[ApiController]
[Route("daily-specials")]
[Produces("application/json")]
[Consumes("application/json")]
public class DailySpecialsController(IDailySpecialsService dailySpecialsService) : ControllerBase
{
    [HttpGet]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public ActionResult<List<DailySpecialResponse>> GetDailySpecials()
    {
        return dailySpecialsService.GetAvailableSpecials();
    }

    [HttpPost("orders")]
    [ProducesResponseType(StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    public async Task<ActionResult<DailySpecialOrderResponse>> OrderDailySpecial(
        [FromBody] DailySpecialOrderRequest request,
        CancellationToken cancellationToken)
    {
        var idempotencyKey = Request.Headers["Idempotency-Key"].FirstOrDefault();

        var cached = await dailySpecialsService.CheckIdempotencyAsync(idempotencyKey, cancellationToken);
        if (cached is not null)
            return StatusCode(StatusCodes.Status201Created, cached);

        var special = dailySpecialsService.ValidateSpecialExists(request.SpecialId);
        if (special is null)
            return NotFound(new ProblemDetails
            {
                Title = "Daily special not found",
                Status = StatusCodes.Status404NotFound,
                Detail = $"No daily special found with ID '{request.SpecialId}'."
            });

        var response = dailySpecialsService.ReserveQuantity(request.SpecialId!.Value, request.Quantity, special.Value.Name);
        if (response is null)
            return Conflict(new ProblemDetails
            {
                Title = "Daily special sold out",
                Status = StatusCodes.Status409Conflict,
                Detail = $"'{special.Value.Name}' has reached the maximum orders for today."
            });

        await dailySpecialsService.StoreIdempotencyResultAsync(idempotencyKey, response, cancellationToken);
        await dailySpecialsService.PublishOrderEventAsync(response, special.Value.Name, cancellationToken);

        return StatusCode(StatusCodes.Status201Created, response);
    }

    [HttpDelete("orders")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public IActionResult ResetOrderCounts([FromQuery] Guid? specialId = null)
    {
        dailySpecialsService.ResetOrderCounts(specialId);
        return NoContent();
    }
}

File: java/breakfast-provider-app/src/main/java/io/lemonlion/breakfast/web/DailySpecialsController.java
package io.lemonlion.breakfast.web;

import io.lemonlion.breakfast.model.request.DailySpecialOrderRequest;
import io.lemonlion.breakfast.model.response.DailySpecialOrderResponse;
import io.lemonlion.breakfast.model.response.DailySpecialResponse;
import io.lemonlion.breakfast.service.DailySpecialsService;
import io.lemonlion.breakfast.service.DailySpecialsService.Special;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ProblemDetail;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

/** Twin of C# {@code DailySpecialsController} ({@code /daily-specials}). */
@RestController
@RequestMapping(path = "/daily-specials", produces = MediaType.APPLICATION_JSON_VALUE)
public class DailySpecialsController {

    private final DailySpecialsService service;
    private final DailySpecialOrderValidator validator;

    public DailySpecialsController(DailySpecialsService service, DailySpecialOrderValidator validator) {
        this.service = service;
        this.validator = validator;
    }

    @GetMapping
    public List<DailySpecialResponse> getDailySpecials() {
        return service.getAvailableSpecials();
    }

    @PostMapping(path = "/orders", consumes = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<?> orderDailySpecial(
            @RequestBody DailySpecialOrderRequest request,
            @RequestHeader(value = "Idempotency-Key", required = false) String idempotencyKey) {

        Optional<DailySpecialOrderResponse> cached = service.checkIdempotency(idempotencyKey);
        if (cached.isPresent()) {
            return ResponseEntity.status(HttpStatus.CREATED).body(cached.get());
        }

        validator.validate(request);

        Optional<Special> special = service.validateSpecialExists(request.specialId());
        if (special.isEmpty()) {
            ProblemDetail problem = ProblemDetail.forStatus(HttpStatus.NOT_FOUND);
            problem.setTitle("Daily special not found");
            problem.setDetail("No daily special found with ID '" + request.specialId() + "'.");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(problem);
        }

        Optional<DailySpecialOrderResponse> response =
                service.reserveQuantity(request.specialId(), request.quantity(), special.get().name());
        if (response.isEmpty()) {
            ProblemDetail problem = ProblemDetail.forStatus(HttpStatus.CONFLICT);
            problem.setTitle("Daily special sold out");
            problem.setDetail("'" + special.get().name() + "' has reached the maximum orders for today.");
            return ResponseEntity.status(HttpStatus.CONFLICT).body(problem);
        }

        service.storeIdempotencyResult(idempotencyKey, response.get());
        service.publishOrderEvent(response.get(), special.get().name());
        return ResponseEntity.status(HttpStatus.CREATED).body(response.get());
    }

    @DeleteMapping("/orders")
    public ResponseEntity<Void> resetOrderCounts(@RequestParam(required = false) UUID specialId) {
        service.resetOrderCounts(specialId);
        return ResponseEntity.noContent().build();
    }
}
